### PR TITLE
Ftr/dev 1306 update with nulls

### PR DESCRIPTION
**Pull request by BrianZito**
**Description:**
Changes nightly uploader to update table values even if they are null. Also refactors the fabs and fpds loaders to reuse more components and be more easily worked on in the future.

**Technical details:**
load_base.py has a store_value function, which has a guard clause that was causing the problem. This was removed. In addition, the upsert_fabs_transactions was converted into a more generic upsert_transactions, and made to be used by both fabs and fpds. fsrs is different enough that refactoring that to work the same way would have been higher risk, and was thus left out.

**Requirements for PR merge:**

1. [x] Unit & integration tests updated
2. [N/A ] API documentation updated
3. [ ] Necessary PR reviewers:
    - [ ] Backend
4. [ ] Matview impact assessment completed
5. [N/A ] Frontend impact assessment completed
6. [N/A ] Data validation completed
7. [N/A ] Appropriate Operations ticket(s) created
8. [ ] Jira Ticket [DEV-1306](https://federal-spending-transparency.atlassian.net/browse/DEV-1306):
    - [x] Link to this Pull-Request
    - [ ] Performance evaluation of affected (API | Script | Download)

**Area for explaining above N/A when needed:**
```
```

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -62,7 +62,7 @@ services:
     volumes:
       - .:/dockermount
     ports:
-      - 8000:8000
+      - 8500:8000
```
Comment (BrianZito) [inline docker-compose.yaml]: Broker backend runs their DB at port 8000, making it impossible to run both at the same time. For future ease, I would advise setting our ports between 8500 and 8799, as broker tends to be all over the place with theirs but avoid that less popular range

File: usaspending_api/broker/helpers/upsert_fabs_transactions.py [hunk 2]
```diff
@@ -218,24 +215,3 @@ def insert_new_fabs(to_insert):
...
-            with timer("updating awards to reflect their latest associated transaction info", logger.info):
-                update_awards(update_award_ids)
-            with timer("updating award category variables", logger.info):
-                update_award_categories(update_award_ids)
-
-        with timer("updating C->D linkages", logger.info):
-            update_c_to_d_linkages("assistance")
-
-    else:
-        logger.info("Nothing to insert...")
```
Comment (BrianZito) [inline usaspending_api/broker/helpers/upsert_fabs_transactions.py]: The biggest part of this PR is renaming this to upsert_transactions and moving it to a common location

File: usaspending_api/etl/management/load_base.py [hunk 3]
```diff
@@ -611,8 +611,6 @@ def get_or_create_location(location_map, row, location_value_map=None, empty_loc
 
 
 def store_value(model_instance_or_dict, field, value, reverse=None):
-    if value is None:
-        return
```
Comment (BrianZito) [inline usaspending_api/etl/management/load_base.py]: This is the actual change in business logic for 1306

File: usaspending_api/common/helpers/etl_helpers.py [hunk 4]
```diff
@@ -59,3 +60,19 @@ def update_c_to_d_linkages(type):
...
+
+    if ids:
+        update_award_ids = tuple(set(ids))  # Convert to tuple and remove duplicates.
+        with timer("updating awards to reflect their latest associated transaction info", logger.info):
+            update_awards(update_award_ids)
+        with timer("updating award category variables", logger.info):
+            update_award_categories(update_award_ids)
+
+        with timer("updating C->D linkages", logger.info):
+            update_c_to_d_linkages(linkage_type)
```
Comment (BrianZito) [inline usaspending_api/common/helpers/etl_helpers.py]: These lines are a little different than in the other version, meaning business logic might have been changed. I can't find any such changes, but absence of evidence is not evidence of absence.

File: docker-compose.yaml (hunk 1 above)
Comment (tony-sappe) [reply docker-compose.yaml]: This isn't a terrible idea. I would suggest that instead of modifying it in this PR we discuss it during a dev meeting to see if USAspending should change or Broker should change.

File: usaspending_api/broker/management/commands/fpds_nightly_loader.py [hunk 5]
```diff
@@ -365,21 +362,12 @@ def handle(self, *args, **options):
             with timer("insertion of new FPDS data in batches", logger.info):
                 self.insert_all_new_fpds(total_insert)
 
-            # Update Awards based on changed FPDS records
-            with timer("updating awards to reflect their latest associated transaction info", logger.info):
-                update_awards(tuple(AWARD_UPDATE_ID_LIST))
-
             # Update FPDS-specific Awards based on the info in child transactions
             with timer("updating contract-specific awards to reflect their latest transaction info", logger.info):
                 update_contract_awards(tuple(AWARD_UPDATE_ID_LIST))
```
Comment (BrianZito) [inline usaspending_api/broker/management/commands/fpds_nightly_loader.py]: In this new version, update_contract_awards now occurs BEFORE update awards. This didn't cause any errors, but it might have changed something via side effects.

Comment (tony-sappe) [reply usaspending_api/broker/management/commands/fpds_nightly_loader.py]: I can't find side-effects as first glance, but don't see a reason that prevents `update_contract_awards()`being after `upsert_transactions()`

File: usaspending_api/broker/management/commands/fabs_nightly_loader.py [hunk 6]
```diff
@@ -200,18 +202,26 @@ def handle(self, *args, **options):
 
         if is_incremental_load and not submission_ids:
             logger.info("No new submissions. Exiting.")
+            return
```
Comment (tony-sappe) [inline usaspending_api/broker/management/commands/fabs_nightly_loader.py]: Adding this return and altering the if statement breaks the functionality of running `update_last_load_date()` when the script is an incremental run and there are no new transactions from Broker

File: usaspending_api/broker/management/commands/load_fsrs.py [hunk 7]
```diff
@@ -499,7 +499,7 @@ def cleanup_broken_links(self, db_cursor):
 
     @transaction.atomic
     def handle(self, *args, **options):
-        logger.info('Starting FSRS data load...')
```
Comment (tony-sappe) [inline usaspending_api/broker/management/commands/load_fsrs.py]: You seen very opinionated on the "correct" log messages

File: usaspending_api/broker/management/commands/fpds_nightly_loader.py (hunk 5 above)
Comment (BrianZito) [reply usaspending_api/broker/management/commands/fpds_nightly_loader.py]: putting it after upsert would also change functionality, since now it's running after the update_transactions and update_c_d_links is called. The other option is to put it in the upsert_transactions() call and use a flag to only have this happen in fpds cases, but that seems un-SOLID and risky in its own way.

File: usaspending_api/etl/management/load_base.py [hunk 8]
```diff
@@ -608,16 +608,20 @@ def get_or_create_location(location_map, row, location_value_map=None, empty_loc
...
-            except (TypeError, ValueError):
-                pass
+    # turn datetimes into dates
+    if field.endswith('date') and isinstance(value, str):
+        try:
+            value = dateutil.parser.parse(value).date()
+        except (TypeError, ValueError):
+            pass
+
+    # handle sorting in reverse order
```
[the remark this answers is not in the document]
Comment (BrianZito) [reply usaspending_api/etl/management/load_base.py]: This was based on what it seemed this code was doing; it would probably be best to just remove the comment.

File: usaspending_api/common/helpers/etl_helpers.py (hunk 4 above)
Comment (BrianZito) [reply usaspending_api/common/helpers/etl_helpers.py]: Do you have a more accurate name to suggest? `upsert_awards`?

File: usaspending_api/broker/management/commands/load_fsrs.py (hunk 7 above)
Comment (BrianZito) [reply usaspending_api/broker/management/commands/load_fsrs.py]: It's less strong opinion and more low cost. These scripts are pretty much doing the same thing, so I feel they should look the same wherever it's cheap to do so.

File: usaspending_api/etl/management/load_base.py [hunk 9]
```diff
@@ -7,7 +7,7 @@
 from datetime import datetime
 from decimal import Decimal
 
-import dateutil
+from dateutil import parser
```
Comment (BrianZito) [inline usaspending_api/etl/management/load_base.py]: After and investigation by kirk and I, it seems that this is the only way to get the store_value method to work with tests.

File: usaspending_api/common/helpers/etl_helpers.py (hunk 4 above)
Comment (kbard) [reply usaspending_api/common/helpers/etl_helpers.py]: Yes, I was a bit confused on the whole `upsert_transactions` accepting award ids as well.  As of right now (about 10% through my code review), I'm for renaming the function.

File: usaspending_api/etl/tests/unit/test_store_value.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    store_value(test_dict, 'test_date_not', good_date)
+    assert test_dict['test_date_not'] == '2005-08-31 00:00:00'
+    assert test_dict['y'] == 6
+
+def test_store_bad_dates():
+    test_dict = {'test_bad_date': 3, 'y': 6}
+    bad_date = [4,7]
+    store_value(test_dict, 'test_date_not', bad_date)
+    assert test_dict['test_date_not'] == [4,7]
+    assert test_dict['y'] == 6
```
Comment (kbard) [inline usaspending_api/etl/tests/unit/test_store_value.py]: A++ for creating tests in a very dark corner of the application!

Any chance I can talk you into cleaning the formatting up a little?  For example, the spacing is wonky on things like `test_dict = {'x':3,'y':6}`.  PyCharm has a code formatting tool that is really nice for cleaning stuff like that up.  Also, certain parties are attempting to get us to standardize on BLACK formatting.  I can help you set that up as well.

File: usaspending_api/broker/management/commands/fabs_nightly_loader.py [hunk 11]
```diff
@@ -198,20 +200,27 @@ def handle(self, *args, **options):
...
 
             with timer("retrieving/diff-ing FABS Data", logger.info):
                 ids_to_upsert = get_fabs_transaction_ids(submission_ids, afa_ids, start_datetime, end_datetime)
 
-            update_award_ids = delete_fabs_transactions(ids_to_delete, do_not_log_deletions)
-            upsert_fabs_transactions(ids_to_upsert, update_award_ids)
+            externally_updated_award_ids = delete_fabs_transactions(ids_to_delete, do_not_log_deletions)
+            
+            if ids_to_upsert or externally_updated_award_ids:
+                update_award_ids = copy(externally_updated_award_ids)
```
[the remark this answers is not in the document]
Comment (BrianZito) [reply usaspending_api/broker/management/commands/fabs_nightly_loader.py]: That call was as-is from before, and Tony's directive was to avoid too unneeded of changes, which makes me reluctant to change something just because it _won't_ make anything bad happen.

File: usaspending_api/broker/management/commands/fabs_nightly_loader.py [hunk 12]
```diff
@@ -198,20 +200,27 @@ def handle(self, *args, **options):
...
+
+                if ids_to_upsert:
+                    with timer("inserting new FABS data", logger.info):
+                        update_award_ids.extend(insert_all_new_fabs(ids_to_upsert))
+
+            upsert_transactions(update_award_ids, "assistance")
```
[the remark this answers is not in the document]
Comment (BrianZito) [reply usaspending_api/broker/management/commands/fabs_nightly_loader.py]: I'm confused what your asking. The if statement at 212 exists to guard this function call; it will always be true, no?

File: usaspending_api/etl/tests/unit/test_store_value.py (hunk 10 above)
Comment (BrianZito) [reply usaspending_api/etl/tests/unit/test_store_value.py]: Now that I look at it, the spacing of those objects isn't even consistent! I'll fix that at least, and I'm happy to hear any formatting standards you're trying to set up.

File: docker-compose.yaml (hunk 1 above)
Comment (kbard) [reply docker-compose.yaml]: A+ for initiative.  Easier to ask for forgiveness...  Plus it got the conversation started.

File: usaspending_api/common/helpers/generic_helper.py [hunk 13]
```diff
@@ -188,7 +188,6 @@ def within_one_year(d1, d2):
             days_diff -= 1
     return days_diff <= 365
 
-
```
Comment (kbard) [inline usaspending_api/common/helpers/generic_helper.py]: This is such a terribly petty thing and I feel absolutely awful for adding this comment, but can we put this blank line back?  My linter is complaining that it wants two spaces between functions and, honestly, this seems like an unnecessary change.

File: usaspending_api/broker/management/commands/fabs_nightly_loader.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (tony-sappe) [reply usaspending_api/broker/management/commands/fabs_nightly_loader.py]: I don't see how I can be blamed for this questionable code you are adding/modifying. :-)

[the remark this answers is not in the document]
Comment (kbard) [reply usaspending_api/broker/management/commands/fabs_nightly_loader.py]: And while avoiding unnecessary edits is an admirable goal, this code was pilfered from another function.  It is yearning, nay, DEMANDING to be refactored in its new environment.

Once refactored, it will highlight another issue that I will bring up in a subsequent review if it is not discovered and fixed beforehand.

File: usaspending_api/broker/management/commands/fabs_nightly_loader.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (kbard) [reply usaspending_api/broker/management/commands/fabs_nightly_loader.py]: My apologies.  This is what I get for reviewing on vacation.  What I MEANT to ask was, what happens here if `if ids_to_upsert or externally_updated_award_ids:` evaluates to *FALSE*.

I get `NameError: name 'update_award_ids' is not defined`.

This is because `upsert_transactions(update_award_ids, "assistance")` is not inside a code block where `update_award_ids` is defined.

File: usaspending_api/common/helpers/etl_helpers.py (hunk 4 above)
Comment (kbard) [inline usaspending_api/common/helpers/etl_helpers.py]: New comment unrelated to the others here... So before, `update_c_to_d_linkages` was called if there were any transaction or award changes.  Now it seems it's being called only for award changes.  Do we know if that's ok?  Looking through the SQL, it doesn't seem like they touch the transaction tables, so this might be fine just want to ensure this was intentional and tested.

File: usaspending_api/broker/management/commands/fabs_nightly_loader.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (BrianZito) [reply usaspending_api/broker/management/commands/fabs_nightly_loader.py]: Ahh, a missing tab. Good eye.

File: usaspending_api/common/helpers/etl_helpers.py (hunk 4 above)
Comment (BrianZito) [reply usaspending_api/common/helpers/etl_helpers.py]: This was noted early on, but the change was more about wanting to keep code clean than desiring an actual business change, and it would probably be best to just move it back. As for testing, since my only way is to test by exhaustion, I've held off on doing a thorough test until the structure of these changes are roughly agreed upon on a style level.

**Comment by BrianZito** [comment]
This isn't getting finished.
